TropicalGeometry: stable_intersection, added perturbation check

YueRen opened the pull request:
Fixes a bug caused by the perturbation chosen for the stable intersection not being checked for genericity (meaning the perturbed polyhedral complexes intersect in too high dimension).  Bug found by @ac2368.

YueRen committed:
TropicalGeometry: stable_intersection, now using lex perturbation

now correct tests whether (random) perturbation is indeed generic. reruns
algorithm if not.

ollieclarke8787 reviewed (approved):
Logic of the lex-perturbation is correct. I agree that using lex-perturbation avoids a requirement to check the genericity of the perturbation. Documentation and example looks correct. Note that quick check 3 will need testing when it becomes available.

YueRen committed:
TropicalGeometry: fixed sign error

YueRen committed:
TropicalGeometry: stable_intersection, added cheap shortcut

Checks whether the intersection is transverse thruogh the already computed
intersection and relative interior point (both necessary for further computations).

YueRen commented:
Just a bump:  Would it be possible to merge this PR?

fingolfin commented:
In triage everyone was happy to merge this. Only question that came up is whether this contains a test for the issue(s) that are being fixed (e.g. does the doctest you add cover this)?

fingolfin commented:
@YueRen if you can reply to the question about tests (and/or add tests) we can proceed with merging this.

YueRen commented:
@fingolfin Sorry for the late reply.  Unfortunately, there is no test for the bug and finding such a test isn't easy either (we found it whilst running a massive experiment, whose data we have since deleted).  

The old method depended on picking a random perturbation and the bug was that the test responsible for checking whether said perturbation was random enough missed some edge-cases.  I tried patching the test, but that only revealed more edge-cases, which is why the current code now uses a symbolic lex perturbation that does not require randomness.  In hindsight, this is also the more efficient approach since perturbing a vector lexicographically is quite fast to do.
